mm2src/crypto/src/hw_ctx.rs

```diff
@@ -55,48 +57,91 @@ impl HardwareWalletCtx {
         let hw_ctx = HardwareWalletArc::new(HardwareWalletCtx {
             hw_internal_pubkey,
-            hw_wallet_type: hw_client.hw_wallet_type(),
-            hw_wallet: AsyncMutex::new(Some(hw_client)),
+            hw_wallet_type: hw_wallet.hw_wallet_type(),
+            hw_wallet,
+            hw_wallet_connected: AtomicBool::new(true),
         });
         Ok((hw_device_info, hw_ctx))
     }
 
     pub fn hw_wallet_type(&self) -> HwWalletType { self.hw_wallet_type }
 
-    /// Connects to a Trezor device and checks if MM was initialized from this particular device.
-    pub async fn trezor<Processor>(
-        &self,
-        processor: &Processor,
-    ) -> MmResult<TrezorClient, HwProcessingError<Processor::Error>>
-    where
-        Processor: TrezorConnectProcessor + Sync,
-        Processor::Error: std::fmt::Display,
-    {
-        let mut hw_client = self.hw_wallet.lock().await;
-        if let Some(HwClient::Trezor(connected_trezor)) = hw_client.deref() {
-            match self.check_trezor(connected_trezor, processor).await {
-                Ok(()) => return Ok(connected_trezor.clone()),
-                // The device could be unplugged. We should try to reconnect to the device.
-                Err(e) => warn!("Error checking hardware wallet device: '{}'. Trying to reconnect...", e),
-            }
+    /// Returns a Trezor session.
+    pub async fn trezor(&self) -> MmResult<TrezorSession<'_>, HwError> {
+        if !self.hw_wallet_connected.load(Ordering::Relaxed) {
+            return MmError::err(HwError::DeviceDisconnected);
         }
-        // Connect to a device.
-        let trezor = HwClient::trezor(processor).await?;
-        // Check if the connected device has the same public key as we used to initialize the app.
-        self.check_trezor(&trezor, processor).await?;
 
-        // Reinitialize the field to avoid reconnecting next time.
-        *hw_client = Some(HwClient::Trezor(trezor.clone()));
+        let HwClient::Trezor(ref trezor) = self.hw_wallet;
+        match trezor.init_new_session().await {
+            Ok((_device, session)) => Ok(session),
+            Err(e) => {
+                self.handle_hw_error(&e);
+                Err(e.map(HwError::from))
+            },
+        }
+    }
+
+    #[cfg(target_arch = "wasm32")]
+    pub async fn trezor_connection_status(&self) -> HwConnectionStatus {
+        if !self.hw_wallet_connected.load(Ordering::Relaxed) {
+            return HwConnectionStatus::Unreachable;
+        }
+
+        let HwClient::Trezor(ref trezor) = self.hw_wallet;
+        let mut session = match trezor.try_session_if_not_occupied() {
+            Some(session) => session,
+            // If we got `None`, the session mutex is occupied by another task,
+            // so for now we can consider the Trezor device as connected.
+            None => return HwConnectionStatus::Connected,
+        };
+
+        #[cfg(target_arch = "wasm32")]
```

Comment:
I believe this is not needed anymore

Reply:
You're right, will remove this compilation flag